src/stream_processor/flb_sp_snapshot.c

```diff
@@ -0,0 +1,275 @@
+/* -*- Mode: C; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+
+/*  Fluent Bit
+ *  ==========
+ *  Copyright (C) 2019      The Fluent Bit Authors
+ *  Copyright (C) 2015-2018 Treasure Data Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <fluent-bit/flb_mem.h>
+#include <fluent-bit/stream_processor/flb_sp.h>
+#include <fluent-bit/stream_processor/flb_sp_parser.h>
+#include <fluent-bit/stream_processor/flb_sp_snapshot.h>
+
+static struct flb_sp_snapshot_page *snapshot_page_create()
+{
+    struct flb_sp_snapshot_page *page;
+
+    page = (struct flb_sp_snapshot_page *)
+           flb_calloc(1, sizeof(struct flb_sp_snapshot_page));
+    if (!page) {
+        flb_errno();
+        return NULL;
+    }
+
+    page->snapshot_page = (char *) flb_malloc(SNAPSHOT_PAGE_SIZE);
+    if (!page->snapshot_page) {
+        flb_errno();
+        flb_free(page);
+        return NULL;
+    }
+
+    return page;
+}
+
+static int snapshot_cleanup(struct flb_sp_snapshot *snapshot, struct flb_time *tms)
+{
+    int ok;
+    size_t off;
```

Comment:
off must be initialized, if I am not wrong this can generate an issue if it gets invalid data when using msgpack iterator in msgpack_unpack_next()